apps/emqx_mix_utils/lib/mix/tasks/compile.asn1.ex

```diff
@@ -0,0 +1,82 @@
+defmodule Mix.Tasks.Compile.Asn1 do
+  use Mix.Task.Compiler
+
+  @recursive true
+  @manifest_vsn 1
+  @manifest "compile.asn1"
+  # TODO: use manifest to track generated files?
+
+  @impl true
+  def manifests(), do: [manifest()]
+  defp manifest(), do: Path.join(Mix.Project.manifest_path(), @manifest)
+
+  @impl true
+  def run(_args) do
+    add_to_path_and_cache(:asn1)
+
+    Mix.Project.get!()
+    config = Mix.Project.config()
+    app_root = File.cwd!()
+
+    asn1_srcs = config[:asn1_srcs] || []
+    asn1_opts = config[:asn1_opts] || %{}
+    out_dir = Map.fetch!(asn1_opts, :out_dir)
+    opts = [:per, :noobj, outdir: to_charlist(out_dir)]
+
+    manifest_data = read_manifest(manifest())
+    manifest_modified_time = Mix.Utils.last_modified(manifest())
+    Enum.each(asn1_srcs, &compile(&1, opts, app_root, manifest_modified_time))
+    write_manifest(manifest(), manifest_data)
+
+    {:noop, []}
+  end
+
+  defp compile(src_path, opts, app_root, manifest_modified_time) do
+    src_path =
+      app_root
+      |> Path.join(src_path)
+      |> Path.expand()
+    if stale?(src_path, manifest_modified_time) do
+      Mix.shell().info("compiling asn1 file: #{src_path}")
+      :ok = :asn1ct.compile(to_charlist(src_path), opts)
+    else
+      Mix.shell().info("file up to date, not compiling: #{src_path}")
```

Comment:
```suggestion
      Mix.shell().info("file is up to date, not compiling: #{src_path}")
```

Reply:
P.S. I am surprised that mix requires writing handling of `is_stale` logic manually. Isn't it bread and butter of build systems?

Reply:
Thanks, fixed.